Copter: position controller updates

lthall opened the pull request:
This is my PR to fix and refactor the position controller. This fixes numerous small bugs and initialization problems and simplifies the interface with the position controller.

amilcarlucas commented:
Hey @lthall , does this already include the position control improvements in guided mode that you talked at in the conference this year?

lthall commented:
> Hey @lthall , does this already include the position control improvements in guided mode that you talked at in the conference this year?

No, but they are coming up behind this one. Want to help me do some testing?

amilcarlucas commented:
Yes, I plan to do some flights this week to test the system identification mode.
Yes, I can help.

rmackay9 commented on libraries/AC_PID/AC_PID_2D.h:
```diff
@@ -36,6 +36,7 @@ class AC_PID_2D {
     Vector2f get_i() const;
     Vector2f get_d() const;
     Vector2f get_ff();
+    Vector2f get_error() const {return _error;}
```
We can make this a const reference and add in some spaces.

const Vector2f& get_error() const { return _error; }

rmackay9 commented on libraries/AC_PID/AC_P_1D.cpp:
```diff
@@ -20,46 +20,60 @@ AC_P_1D::AC_P_1D(float initial_p, float dt) :
 
     _kp = initial_p;
-    _lim_D_Out = 10.0f;     // maximum first differential of output
+    _error_min = 0.0f;
```
No need to set these values to zero.  They will be zero by default I think.  Right @peterbarker?

lthall replied:
@peterbarker Can you confirm?

peterbarker replied:
Yes.  Unless you're ever constructing one of these on the stack you do not need to initialise them.

Here's the summary: https://ardupilot.org/dev/docs/style-guide.html#implicit-zeroing-of-memory

lthall replied:
Fixed in a few placed in the PID libary.

rmackay9 commented on libraries/AC_PID/AC_P_1D.cpp:
```diff
@@ -20,46 +20,60 @@ AC_P_1D::AC_P_1D(float initial_p, float dt) :
+    if (is_positive(D2_max)) {
         // limit the first derivative so as not to exceed the second derivative
-        _lim_D_Out = MIN(_lim_D_Out, lim_D2_Out / _kp);
+        _D1_max = MIN(_D1_max, D2_max / _kp);
```
I'm slightly surprised we don't have a is_positive() check here for _kp.

rmackay9 commented on libraries/AC_PID/AC_P_1D.cpp:
```diff
@@ -20,46 +20,60 @@ AC_P_1D::AC_P_1D(float initial_p, float dt) :
-    _lim_err_min = MAX(inv_sqrt_controller(lim_out_min, _kp, _lim_D_Out), lim_err_min);
-    _lim_err_max = MAX(inv_sqrt_controller(lim_out_max, _kp, _lim_D_Out), lim_err_max);
+    _error_min = inv_sqrt_controller(output_min, _kp, _D1_max);
+    if(!is_zero(_error_min) && is_negative(error_min) ) {
```
nitpick: we should have a space between "if" and the following bracket.  Same thing a few lines lower.

rmackay9 commented on libraries/AC_PID/AC_P_1D.h:
```diff
@@ -17,22 +17,29 @@ class AC_P_1D {
     void set_dt(float dt) { _dt = dt; }
 
+    // get maximum error
+    float get_error_min() const {return _error_min;}
```
nitpick: we should have a space after/before the curly brackets.  same thing on the line below.

rmackay9 commented on ArduCopter/mode_althold.cpp:
```diff
@@ -75,8 +73,7 @@ void ModeAltHold::run()
         // set position controller targets
-        pos_control->set_alt_target_from_climb_rate_ff(target_climb_rate, G_Dt, false);
-        pos_control->add_takeoff_climb_rate(takeoff_climb_rate, G_Dt);
+        pos_control->set_alt_target_from_climb_rate(target_climb_rate + takeoff_climb_rate, false);
```
I would have expected "set_alt_target_from_climb_rate" to be renamed to, "set_pos_target_z_from_climb_rate"

rmackay9 commented on libraries/AC_AttitudeControl/AC_PosControl.cpp:
```diff
@@ -81,7 +81,7 @@ const AP_Param::GroupInfo AC_PosControl::var_info[] = {
     // @Increment: 0.1
     // @User: Advanced
-    AP_GROUPINFO("_ACC_XY_FILT", 1, AC_PosControl, _accel_xy_filt_hz, POSCONTROL_ACCEL_FILTER_HZ),
+    // AP_GROUPINFO("_ACC_XY_FILT", 1, AC_PosControl, _accel_xy_filt_hz, POSCONTROL_ACCEL_FILTER_HZ),
```
we normally don't leave commented out code in place so if it's gone, let's remove this _accel_xy_filt_hz variable completely.

rmackay9 commented:
I think we should consider splitting out the rename and formatting changes into a separate PR for easier review.  This is a list of the direct one-to-one renames I've found.

pos_control->get_alt_error TO pos_control->get_pos_error_z_cm
pos_control->get_bearing_to_target TO pos_control->get_bearing_to_target_cd
pos_control->get_desired_velocity TO pos_control->set_vel_desired_xy_cms
pos_control->get_pos_error_xy() TO pos_control->get_pos_error_xy_cm
pos_control->get_roll TO pos_control->get_roll_cd
pos_control->get_pitch TO pos_control->get_pitch_cd
pos_control->get_max_accel_xy TO pos_control->get_max_accel_xy_cmss
pos_control->get_max_accel_z TO pos_control->get_max_accel_z_cmss
pos_control->get_desired_velocity TO pos_control->get_vel_desired_cms
pos_control->set_alt_target to pos_control->set_pos_target_z_cm
pos_control->set_desired_accel_xy TO pos_control->set_accel_desired_xy_cmss
pos_control->set_desired_velocity_xy TO pos_control->set_vel_desired_xy_cms
pos_control->set_desired_velocity_z TO pos_control->set_vel_desired_z_cms

rmackay9 commented on ArduSub/Sub.cpp:
```diff
@@ -34,7 +34,7 @@ Sub::Sub()
           ahrs_view(ahrs, ROTATION_NONE),
           attitude_control(ahrs_view, aparm, motors, MAIN_LOOP_SECONDS),
-          pos_control(ahrs_view, inertial_nav, motors, attitude_control),
+          pos_control(ahrs_view, inertial_nav, motors, attitude_control, MAIN_LOOP_SECONDS),
```
although the call to the attitude controller directly above it also uses MAIN_LOOP_SECONDS, sure this should use, "scheduler.get_loop_period_s()" shouldn't it?

rmackay9 commented on libraries/AC_AttitudeControl/AC_PosControl.cpp:
```diff
@@ -335,278 +343,566 @@ void AC_PosControl::set_dt(float delta_sec)
-    _accel_z_cms = accel_cmss;
-    _flags.recalc_leash_z = true;
-    calc_leash_length_z();
+    if ( is_positive(accel_limit_cmss)) {
```
nitpick: little extra space after the first bracket

rmackay9 commented on libraries/AC_AttitudeControl/AC_PosControl.cpp:
```diff
@@ -335,278 +343,566 @@ void AC_PosControl::set_dt(float delta_sec)
+
+    // ensure the horizontal time constant is not less than the vehicle is capable of
+    float lean_angle = _accel_max_xy_cmss / (GRAVITY_MSS * 100.0 * M_PI / 18000.0);
+    float angle_accel = MIN(_attitude_control.get_accel_pitch_max(), _attitude_control.get_accel_roll_max());
```
we probably need to protect against divide-by-zero here because I think _attitude_control.get_accel_pitch_max() or _attitude_control.get_accel_roll_max() could return zero.

rmackay9 commented on libraries/AC_AttitudeControl/AC_PosControl.h:
```diff
@@ -29,265 +27,279 @@
 
 #define POSCONTROL_OVERSPEED_GAIN_Z             2.0f    // gain controlling rate at which z-axis speed is brought back within SPEED_UP and SPEED_DOWN range
 
+#define POSCONTROL_RELAX_TC                     0.16f
```
I guess all the other definitions get comments so this one should too

lthall replied:
Yes, it felt left out.

rmackay9 commented on libraries/AC_AttitudeControl/AC_PosControl.h:
```diff
@@ -29,265 +27,279 @@
-    ///     target will also be stopped if the motors hit their limits or leash length is exceeded
-    ///     set force_descend to true during landing to allow target to move low enough to slow the motors
-    virtual void set_alt_target_from_climb_rate(float climb_rate_cms, float dt, bool force_descend);
+    void set_shaper_tc_xy(float tc_xy_s) {_tc_xy_s = MAX(_tc_xy_s, tc_xy_s);}
```
looks like set_shaper_tc_xy() is never called so it should probably be removed or at least get a comment if we know it will be used in the near future.

rmackay9 commented on libraries/AC_AttitudeControl/AC_PosControl.h:
```diff
@@ -29,265 +27,279 @@
-    /// get_vel_z_error_ratio - returns the proportion of error relative to the maximum request
-    float get_vel_z_control_ratio() const { return constrain_float(_vel_z_control_ratio, 0.0f, 1.0f); }
+    /// stop_pos_xy_stabilisation
+    void stop_pos_xy_stabilisation();
```
maybe stop_pos_xy_stabilisation() is never called?

lthall replied:
Used in guided and plane pr's

rmackay9 commented on libraries/AC_AttitudeControl/AC_PosControl.h:
```diff
@@ -29,265 +27,279 @@
-    // overrides the velocity process variable for one timestep
-    void override_vehicle_velocity_xy(const Vector2f& vel_xy) { _vehicle_horiz_vel = vel_xy; _flags.vehicle_horiz_vel_override = true; }
+    // get_vel_target_cms - returns the target velocity in NEU cm/s
+    const Vector3f& get_vel_error_cms() const { return _vel_error; }
```
get_vel_error_cms() is also never called it seems

lthall replied:
Used in Plane. I checked these trying to delete any we don't use in the following guided or plane patches.

rmackay9 commented on libraries/AC_WPNav/AC_Loiter.cpp:
```diff
@@ -309,12 +282,19 @@ void AC_Loiter::calc_desired_velocity(float nav_dt, bool avoidance_on)
 
+    // get loiters desired velocity from the position controller where it is being stored.
+    const Vector3f &target_pos_3d = _pos_control.get_pos_target_cm();
+    Vector2f target_pos(target_pos_3d.x,target_pos_3d.y);
```
I'm sure PeterB would prefer we use curly braces for this line.  So make it, "Vector2f target_pos{target_pos_3d.x,target_pos_3d.y};".

rmackay9 commented on libraries/AP_Math/control.cpp:
```diff
@@ -22,18 +22,203 @@
+void update_vel_accel(float& vel, float accel, float dt,
+    bool limit_neg, bool limit_pos, float vel_error)
+{
+    float delta_vel = accel * dt;
```
nitpick: we can put a const before, "float delta_vel"

lthall replied:
done

rmackay9 commented on libraries/AP_Math/control.cpp:
```diff
@@ -22,18 +22,203 @@
+    // increase velocity by acceleration * dt if it does not increase error when limited.
+    Vector2f delta_vel = accel * dt;
+
+    if(limit && !is_zero(vel_error.length_squared())) {
```
nitpick: space after "if"

lthall replied:
fixed

rmackay9 commented on libraries/AP_Math/control.cpp:
```diff
@@ -22,18 +22,203 @@
-    vel = vel + accel * dt;
+    Vector2f delta_pos = vel * dt + accel * 0.5f * sq(dt);
+
+    if(limit && !is_zero(pos_error.length_squared())) {
```
space after "if"

lthall commented on ArduCopter/system.cpp:
```diff
@@ -544,7 +542,7 @@ void Copter::allocate_motors(void)
     AP_Param::load_object_from_eeprom(attitude_control, ac_var_info);
         
-    pos_control = new AC_PosControl(*ahrs_view, inertial_nav, *motors, *attitude_control);
+    pos_control = new AC_PosControl(*ahrs_view, inertial_nav, *motors, *attitude_control, scheduler.get_loop_period_s());
```
I have removed it but we can add it back if needed.

lthall commented on libraries/AC_AttitudeControl/AC_PosControl.cpp:
```diff
@@ -748,88 +1034,37 @@ void AC_PosControl::get_stopping_point_xy(Vector3f &stopping_point) const
     // convert the stopping distance into a stopping point using velocity vector
-    stopping_point.x = curr_pos.x + (stopping_dist * curr_vel.x / vel_total);
-    stopping_point.y = curr_pos.y + (stopping_dist * curr_vel.y / vel_total);
+    if ( is_positive(stopping_dist) && is_positive(vel_total)) {
```
got it

lthall commented on libraries/AP_Math/control.cpp:
```diff
@@ -22,18 +22,203 @@
+void update_vel_accel_xy(Vector3f& vel, const Vector3f& accel, float dt,
+    bool limit, Vector2f vel_error)
+{
+    Vector2f vel_2f = Vector2f(vel.x, vel.y);
```
Fixed in all directories I am changing.

lthall commented:
I was also wondering if we should change "set_alt_target_with_slew"

rmackay9 commented on libraries/AC_PID/AC_P_1D.h:
```diff
@@ -17,22 +17,29 @@ class AC_P_1D {
 
+    // set limits on error, output and output from D term
+    void set_limits(float error_min, float error_max, float output_min, float output_max, float D1_max = 0.0f, float D2_max = 0.0f);
+    float get_error_min() { return _error_min; }
```
Looks like we have duplicate calls called get_error_min() and get_error_max().  This is possible because one of them is "const" and one is not.  We should remove this one which isn't const.

lthall replied:
That hit my blind spot! I had to do a search to find the other ones and it was only a couple lines above!

rmackay9 commented on libraries/AC_WPNav/AC_Circle.cpp:
```diff
@@ -222,17 +205,18 @@ bool AC_Circle::update()
         // heading is same as _angle but converted to centi-degrees
         _yaw = _angle * DEGX100;
     }
+    pos.z = target_z_cm;
```
We could change this line to, "pos.z = _center.z + terr_offset" and then remove this section a bit higher up

```
    // calculate z-axis target
    float target_z_cm;
    if (_terrain_alt) {
        target_z_cm = _center.z + terr_offset;
    } else {
        target_z_cm = _center.z;
    }
```

lthall replied:
done

rmackay9 commented on libraries/AC_AttitudeControl/AC_PosControl.cpp:
```diff
@@ -81,7 +81,6 @@ const AP_Param::GroupInfo AC_PosControl::var_info[] = {
     // @Range: 0.5 5
```
I think we can remove this parameter meta data for the PSC_ACC_XY_FILT parameter as well.

lthall replied:
Done

rmackay9 commented on ArduCopter/mode_guided.cpp:
```diff
@@ -162,15 +162,14 @@ void ModeGuided::vel_control_start()
     // initialize vertical speeds and acceleration
-    pos_control->set_max_speed_z(-get_pilot_speed_dn(), g.pilot_speed_up);
-    pos_control->set_max_accel_z(g.pilot_accel_z);
+    pos_control->set_max_speed_accel_z(wp_nav->get_default_speed_down(), wp_nav->get_default_speed_up(), wp_nav->get_accel_z());
```
Guided mode's velocity controller changes behaviour here slightly in that we now use the wpnav speed up and speed down (also note the missing negative).  I think it makes more sense to use the wpnav limits but it is a change.

lthall replied:
Good point but I agree that is should be based on WPNav.

set_max_speed_accel_z will accept positive or negative values for speed down. speed_down = -fabsf(speed_down);

lthall commented:
fixed, squashed, pushed.

rmackay9 commented on ArduCopter/mode_throw.cpp:
```diff
@@ -19,6 +19,9 @@ bool ModeThrow::init(bool ignore_checks)
     stage = Throw_Disarmed;
     nextmode_attempted = false;
 
+    // initialise the z controller
```
It does no harm (besides consuming a bit of flash) but I don't think we need these upper four calls to init_z_controller().  The position controller is only run in Throw_HgtStabilise and Throw_PosHold and line 73 already has the init_z_controller_stopping_point().

lthall replied:
Removed them.

tridge commented:
@lthall looks good so far to me. I have fixed up the calls for quadplane here:
https://github.com/tridge/ardupilot/tree/plane4.1
that is based on this PR

rmackay9 commented:
I have tested this quite extensively in SITL over the past two days and Leonard has resolved all the known issues.

Here are the three known changes (mostly improves) vs latest (aka "master"):

- Throw modes stops much more aggressively which is probably a good thing.  It is now consistent with Brake mode
- Guided mode velocity controllers uses the waypoint nav speed and acceleration limits (e.g. WPNAV_SPEED, WPNAV_ACCEL, WPNAV_SPEED_UP, DN) instead of the "pilot" speed limits used for Loiter.  This also makes more sense than "latest"
- Takeoff in Loiter, AltHold, PosHold is slightly more aggressive than "latest" when PILOT_TKOFF_ALT is non-zero

Modes tested so far: AltHold, Brake, Circle, Flip, Guided (velocity only), Land, Loiter, Throw
Modes still to be tested: Auto, AutoTune, FlowHold, Follow, Guided (pos-vel, attitude), PosHold, RTL, SmartRTL, Sport, ZigZag

Although it may appear less than 1/2 the modes have been tested, there is quite a lot in common between modes so I suspect the remaining testing will go quickly.

lthall commented:
I have simplified the multidimensional limit handling and added some features to support quadplane. All effectively non-functional changes focused on making it easier to understand and interact with the algorithms.

tridge commented:
there is an issue with non lock-step simulators we need to fix, with the panic in pos controller

tridge commented:
> there is an issue with non lock-step simulators we need to fix, with the panic in pos controller

i've fixed this now

peterbarker commented on libraries/AP_Math/control.cpp:
```diff
@@ -118,16 +355,91 @@ void shape_pos_vel(float& pos_input, float vel_input, float pos, float vel, floa
+    const Vector3f& pos, const Vector3f& vel, Vector3f& accel,
+    float vel_max, float vel_correction_max, float accel_max, float tc, float dt)
+{
+    Vector2f pos_input_2f = Vector2f{pos_input.x, pos_input.y};
+    Vector2f vel_input_2f = Vector2f{vel_input.x, vel_input.y};
+    Vector2f accel_input_2f = Vector2f{accel_input.x, accel_input.y};
+    Vector2f pos_2f = Vector2f{pos.x, pos.y};
+    Vector2f vel_2f = Vector2f{vel.x, vel.y};
+    Vector2f accel_2f = Vector2f{accel.x, accel.y};
```
```suggestion
    const Vector2f pos_input_2f {pos_input.x, pos_input.y};
    const Vector2f vel_input_2f {vel_input.x, vel_input.y};
    const Vector2f accel_input_2f {accel_input.x, accel_input.y};
    const Vector2f pos_2f {pos.x, pos.y};
    const Vector2f vel_2f {vel.x, vel.y};
    Vector2f accel_2f {accel.x, accel.y};
```

peterbarker commented on libraries/AP_Math/control.cpp:
```diff
@@ -180,59 +492,33 @@ Vector2f sqrt_controller(const Vector2f& error, float p, float second_ord_lim, f
 // inverse of the sqrt controller.  calculates the input (aka error) to the sqrt_controller required to achieve a given output
 float inv_sqrt_controller(float output, float p, float D_max)
```
We really should have some unit tests for this function.  We can probably test it against the sqrt_controller, after all.

lthall replied:
I don't know how to write ro run unit tests.

peterbarker reviewed (changes requested):
Commit messages need help; "fixes" really isn't adequate.  An outline of the macro-level changes in the extended commit message would be good.

Method renames should have been pulled out as a separate set of commits on which this was based.

Why time-constants in some place and Hz in others?

The massive comments above the methods are great - but I think they need a lint for copy/paste errors.

Sorry, review got a bit shaky going through control.cpp - will try to revisit that one.

lthall commented on libraries/AC_PID/AC_PID_2D.cpp:
```diff
@@ -121,24 +121,49 @@ Vector2f AC_PID_2D::update_all(const Vector2f &target, const Vector2f &measureme
+Vector2f AC_PID_2D::update_all(const Vector3f &target, const Vector3f &measurement, const Vector3f &limit)
 {
-    return update_all(Vector2f{target.x, target.y}, Vector2f{measurement.x, measurement.y}, limit);
+    return update_all(Vector2f{target.x, target.y}, Vector2f{measurement.x, measurement.y}, Vector2f{limit.x, limit.y});
```
Yes, I would like to do this better but I figured the first step is to get all the math right. Once we have this consolidated I hope we can go through and optimise the way it is coded.

lthall commented on ArduCopter/Log.cpp:
```diff
@@ -35,8 +35,8 @@ void Copter::Log_Write_Control_Tuning()
     if (!flightmode->has_manual_throttle()) {
-        des_alt_m = pos_control->get_alt_target() / 100.0f;
-        target_climb_rate_cms = pos_control->get_vel_target_z();
+        des_alt_m = pos_control->get_pos_target_z_cm() / 100.0f;
```
@rmackay9 Do you want me to do this and if so what convention do you want me to use?

lthall replied:
Randy says no more name changes.

lthall commented on libraries/AP_Math/control.cpp:
```diff
@@ -118,16 +355,91 @@ void shape_pos_vel(float& pos_input, float vel_input, float pos, float vel, floa
+    Vector2f vel_2f = Vector2f{vel.x, vel.y};
+    Vector2f accel_2f = Vector2f{accel.x, accel.y};
+
+    shape_pos_vel_accel_xy(pos_input_2f, vel_input_2f, accel_input_2f,
+        pos_2f, vel_2f, accel_2f, vel_max, vel_correction_max, accel_max, tc, dt);
```
Do you really want to write them all like this?

peterbarker replied:
Maybe not for this PR - but I kind of like the style.

lthall replied:
cool

lthall commented on ArduCopter/mode_althold.cpp:
```diff
@@ -68,15 +64,11 @@ void ModeAltHold::run()
-        // set position controller targets
-        pos_control->set_alt_target_from_climb_rate_ff(target_climb_rate, G_Dt, false);
-        pos_control->add_takeoff_climb_rate(takeoff_climb_rate, G_Dt);
+        // get take-off adjusted pilot and takeoff climb rates
```
// run pilot take off procedure

lthall commented on ArduCopter/mode_brake.cpp:
```diff
@@ -9,17 +9,18 @@
+    pos_control->set_max_speed_accel_xy(inertial_nav.get_velocity().length(), BRAKE_MODE_DECEL_RATE);
+
+    // initialise position controller
+    pos_control->init_xy_controller();
```
We may not need to init the xy position controller each time but this is the safe option because there are a few different ways we work with the horizontal axis while the z axis is more consistent.

As this consolidates I will be looking to see if I should or should not init every time.

lthall commented on ArduCopter/mode_guided.cpp:
```diff
@@ -179,22 +178,15 @@ void ModeGuided::posvel_control_start()
-    // set vertical speed and acceleration
-    pos_control->set_max_speed_z(wp_nav->get_default_speed_down(), wp_nav->get_default_speed_up());
-    pos_control->set_max_accel_z(wp_nav->get_accel_z());
+    // initialise velocity controller
+    pos_control->init_z_controller();
+    pos_control->init_xy_controller();
```
I agree. As this is a 2.5D controller I would do it the way we have done here. It will take time to get all the use cases set up properly as there are some structural issues that will need to be changed.

lthall commented on ArduCopter/mode_loiter.cpp:
```diff
@@ -131,22 +127,17 @@ void ModeLoiter::run()
         // get avoidance adjusted climb rate
         target_climb_rate = get_avoidance_adjusted_climbrate(target_climb_rate);
 
+        // get take-off adjusted pilot and takeoff climb rates
```
I like your comment better :)

lthall commented on ArduCopter/mode_loiter.cpp:
```diff
@@ -131,22 +127,17 @@ void ModeLoiter::run()
         target_climb_rate = get_avoidance_adjusted_climbrate(target_climb_rate);
 
+        // get take-off adjusted pilot and takeoff climb rates
+        takeoff.do_pilot_takeoff(target_climb_rate);
```
No, I don't think that is clear at all.

lthall commented on ArduCopter/mode_loiter.cpp:
```diff
@@ -157,8 +148,7 @@ void ModeLoiter::run()
         attitude_control->input_euler_angle_roll_pitch_euler_rate_yaw(0.0f, 0.0f, 0.0f);
-        pos_control->relax_alt_hold_controllers(0.0f);   // forces throttle output to go to zero
-        pos_control->update_z_controller();
+        pos_control->relax_z_controller(0.0f);   // forces throttle output to go to zero
```
I used: forces throttle output to decay to zero

lthall commented on ArduCopter/mode_loiter.cpp:
```diff
@@ -183,10 +173,12 @@ void ModeLoiter::run()
         break;
     }
+
+    // call z-axis position controller
```
cleaned this up with these two comments depending on location:
    // WP_Nav has set the vertical position control targets
    // run the vertical position controller and set output throttle

lthall commented on ArduCopter/takeoff.cpp:
```diff
@@ -48,95 +48,50 @@ bool Mode::do_user_takeoff(float takeoff_alt_cm, bool must_navigate)
 }
 
-// start takeoff to specified altitude above home in centimeters
+// start takeoff to specified altitude above home in centimetres
```
done

lthall commented on ArduCopter/takeoff.cpp:
```diff
@@ -48,95 +48,50 @@ bool Mode::do_user_takeoff(float takeoff_alt_cm, bool must_navigate)
-    start_ms = millis();
-    alt_delta = alt_cm;
+    take_off_start_alt = copter.pos_control->get_pos_target_z_cm();
+    take_off_alt = take_off_start_alt + alt_cm;
```
take_off_complete_alt

lthall commented on ArduCopter/takeoff.cpp:
```diff
@@ -48,95 +48,50 @@ bool Mode::do_user_takeoff(float takeoff_alt_cm, bool must_navigate)
-//  takeoff_climb_rate is only an output
-//  has side-effect of turning takeoff off when timeout as expired
-void Mode::_TakeOff::get_climb_rates(float& pilot_climb_rate, float& takeoff_climb_rate)
+// do_pilot_takeoff - commands the aircraft to the take off altitude
```
// do_pilot_takeoff - controls the vertical position controller during the process of taking off
//  take off is complete when the vertical target reaches the take off altitude.
//  climb is cancelled if pilot_climb_rate_cm becomes negative
//  sets take off to complete when target altitude is within 1% of the take off altitude

lthall commented on ArduCopter/takeoff.cpp:
```diff
@@ -48,95 +48,50 @@ bool Mode::do_user_takeoff(float takeoff_alt_cm, bool must_navigate)
 
-    const float time_to_max_speed = (max_speed - takeoff_minspeed) / TAKEOFF_ACCEL;
-    float height_gained;
-    if (time_elapsed <= time_to_max_speed) {
-        height_gained = 0.5f * TAKEOFF_ACCEL * sq(time_elapsed) + takeoff_minspeed * time_elapsed;
-    } else {
-        height_gained = 0.5f * TAKEOFF_ACCEL * sq(time_to_max_speed) + takeoff_minspeed * time_to_max_speed +
-                        (time_elapsed - time_to_max_speed) * max_speed;
-    }
```
Because people hack shit in without understanding what they are doing. This has been done all over the place and is one of the main reasons this rewrite was so difficult.

In all fairness what they did worked so I can respect that :)

lthall replied:
At the time this did change the existing behaviour however.

lthall commented on ArduCopter/takeoff.cpp:
```diff
@@ -48,95 +48,50 @@ bool Mode::do_user_takeoff(float takeoff_alt_cm, bool must_navigate)
-        } else {
-            pilot_climb_rate = 0.0f;
-        }
+    // stop take off early and return if negative climb rate is commanded or we are within 0.1% of our take off altitude
```
This is because the target altitude decays to the take off altitude. Therefore it will never reach this altitude and instead just get closer and closer. We would then need to set wait for a pilot input to leave the take off state. Randy prefered this check.

lthall commented on ArduPlane/qautotune.cpp:
```diff
@@ -41,8 +41,9 @@ void QAutoTune::get_pilot_desired_rp_yrate_cd(float &des_roll_cd, float &des_pit
 {
-    plane.quadplane.pos_control->set_max_speed_z(-plane.quadplane.get_pilot_velocity_z_max_dn(), plane.quadplane.pilot_velocity_z_max_up);
-    plane.quadplane.pos_control->set_max_accel_z(plane.quadplane.pilot_accel_z);
+    plane.quadplane.pos_control->set_max_speed_accel_z(-plane.quadplane.get_pilot_velocity_z_max_dn(),
```
Randy said no more name changes. The units are in the parameter names so that is probably good enough.
void set_max_speed_accel_xy(float speed_cms, float accel_cmss, float accel_limit_cmss = 0.0f);

lthall commented on ArduPlane/quadplane.cpp:
```diff
@@ -728,7 +728,6 @@ bool QuadPlane::setup(void)
     default:
         AP_BoardConfig::config_error("Unsupported Q_FRAME_CLASS %u", frame_class);
     }
-    ::printf("QuadPlane initialise, class: %u, type: %u\n", frame_class.get(), frame_type.get());
```
This is one for @tridge

tridge replied:
it ended up printing twice, and only on console, which isn't useful for most people (needs to be mavlink to be visible in missionplanner)

lthall commented on ArduPlane/quadplane.cpp:
```diff
@@ -779,7 +778,7 @@ bool QuadPlane::setup(void)
 
     AP_Param::load_object_from_eeprom(attitude_control, attitude_control->var_info);
-    pos_control = new AC_PosControl(*ahrs_view, inertial_nav, *motors, *attitude_control);
+    pos_control = new AC_PosControl(*ahrs_view, inertial_nav, *motors, *attitude_control, loop_delta_t);
```
Yep, leave that for the coding gods

lthall commented on ArduPlane/quadplane.cpp:
```diff
@@ -1081,27 +1079,18 @@ void QuadPlane::control_stabilize(void)
-            if (pos_control->get_alt_target() < inertial_nav.get_altitude()) {
-                set_alt_target_current();
-            }
+            // initialize the Alt Hold controller with no decent
```
This spelling mistake is made in lots of placed. I have fixed the ones in the files I am touching.

lthall commented on ArduPlane/quadplane.cpp:
```diff
@@ -1247,6 +1238,24 @@ void QuadPlane::get_pilot_desired_lean_angles(float &roll_out_cd, float &pitch_o
+*/
+float QuadPlane::get_pilot_land_throttle(void) const
+{
+    if (plane.rc_failsafe_active()) {
```
One for @tridge

tridge replied:
it is good as it is

lthall commented on ArduSub/control_guided.cpp:
```diff
@@ -229,7 +217,8 @@ bool Sub::guided_set_destination_posvel(const Vector3f& destination, const Vecto
 
-    pos_control.set_pos_target(posvel_pos_target_cm);
+    pos_control.input_pos_vel_accel_xy(posvel_pos_target_cm, posvel_vel_target_cms, Vector3f());
+    pos_control.input_pos_vel_accel_z(posvel_pos_target_cm, posvel_vel_target_cms, Vector3f());
```
Yes, it should not have been. If it is set to zero then it will be ignored.

lthall commented on ArduSub/control_guided.cpp:
```diff
@@ -469,7 +457,8 @@ void Sub::guided_angle_control_run()
         attitude_control.relax_attitude_controllers();
-        pos_control.relax_alt_hold_controllers(motors.get_throttle_hover());
+        // initialise velocity controller
+        pos_control.init_z_controller();
```
Because it isn't armed.

rmackay9 commented:
An update on my testing:

Modes tested so far: AltHold, Auto, Brake, Circle, Flip, Guided (velocity only), Land, Loiter, PosHold, RTL, SmartRTL, Sport, Throw, ZigZag
Modes still to be tested: AutoTune, FlowHold, Follow, Guided (pos-vel, attitude),

tridge commented on ArduPlane/quadplane.cpp:
```diff
@@ -2455,6 +2496,53 @@ bool QuadPlane::in_vtol_posvel_mode(void) const
+{
+    if ((options & OPTION_THR_LANDING_CONTROL) == 0) {
+        // not enabled
+        poscontrol.pilot_correction_active = false;
```
its part of the state for ship landing, I'd like to leave it in to make those patches easier

tridge commented on ArduPlane/quadplane.cpp:
```diff
@@ -2455,6 +2496,53 @@ bool QuadPlane::in_vtol_posvel_mode(void) const
+
+    poscontrol.pilot_correction_active = (fabsf(roll_in) > 0 || fabsf(pitch_in) > 0);
+    if (poscontrol.pilot_correction_active) {
+        poscontrol.pilot_correction_done = true;
```
it's fine

tridge commented on ArduPlane/quadplane.cpp:
```diff
@@ -2564,37 +2652,59 @@ void QuadPlane::vtol_position_controller(void)
+
+        run_xy_controller();
+
+        // nav roll and pitch are controller by position controller
```
fix later

lthall commented on Tools/autotest/arducopter.py:
```diff
@@ -5065,7 +5065,7 @@ def fly_poshold_takeoff(self):
             # make sure we haven't already reached alt:
             m = self.mav.recv_match(type='GLOBAL_POSITION_INT', blocking=True)
-            max_initial_alt = 500
+            max_initial_alt = 2000
```
I have asked bill to check this for heli but I don't think it is a problem for copter. As you say we are going to highlight it in the release notes.

lthall commented on libraries/AC_AttitudeControl/AC_PosControl.cpp:
```diff
@@ -279,333 +296,594 @@ const AP_Param::GroupInfo AC_PosControl::var_info[] = {
+///     The time constant defines the acceleration error decay in the kinematic path as the system approaches constant acceleration.
+///     The time constant also defines the time taken to achieve the maximum acceleration.
+///     The function alters the input velocity to be the velocity that the system could reach zero acceleration in the minimum time.
+void AC_PosControl::input_pos_vel_accel_xyz(Vector3f& pos)
```
It will be in guided pr.

lthall commented on libraries/AC_AttitudeControl/AC_PosControl.cpp:
```diff
@@ -279,333 +296,594 @@ const AP_Param::GroupInfo AC_PosControl::var_info[] = {
+///
+
+/// set_max_speed_accel_xy - set the maximum horizontal speed in cm/s and acceleration in cm/s/s
+void AC_PosControl::set_max_speed_accel_xy(float speed_cms, float accel_cmss, float accel_limit_cmss)
```
units in parameter inputs and Randy says no more name changes.

lthall commented on libraries/AC_AttitudeControl/AC_PosControl.cpp:
```diff
@@ -279,333 +296,594 @@ const AP_Param::GroupInfo AC_PosControl::var_info[] = {
+
+    // ensure the horizontal time constant is not less than the vehicle is capable of
+    float lean_angle = _accel_max_xy_cmss / (GRAVITY_MSS * 100.0 * M_PI / 18000.0);
+    float angle_accel = MIN(_attitude_control.get_accel_pitch_max(), _attitude_control.get_accel_roll_max());
```
No, one of these is the minimum of the elipse.

lthall commented on libraries/AC_AttitudeControl/AC_PosControl.cpp:
```diff
@@ -634,109 +912,108 @@ void AC_PosControl::run_z_controller()
     // Check for vertical controller health
 
     // _speed_down_cms is checked to be non-zero when set
-    float error_ratio = _vel_error.z/_speed_down_cms;
+    float error_ratio = _vel_error.z/_vel_max_down_cms;
```
Because we only care about coming down when we hare having a fly away.

lthall commented on libraries/AC_AttitudeControl/AC_PosControl.cpp:
```diff
@@ -1074,31 +1107,20 @@ void AC_PosControl::accel_to_lean_angles(float accel_x_cmss, float accel_y_cmss,
-    Vector3f accel_target = get_accel_target();
-    accel_target.z = -GRAVITY_MSS * 100.0f;
-    return accel_target;
+    accel_x_cmss = accel_cmss.x;
+    accel_y_cmss = accel_cmss.y;
```
This is not as simple as that.

lthall commented on libraries/AC_AttitudeControl/AC_PosControl.h:
```diff
@@ -29,245 +27,253 @@
-    /// set_alt_target_to_current_alt - set altitude target to current altitude
-    void set_alt_target_to_current_alt() { _pos_target.z = _inav.get_altitude(); }
+    // set the maximum horizontal position error that will be allowed in the horizontal plane
+    void set_pos_error_max_xy_cm(float error_max) { _p_pos_xy.set_error_max(error_max); }
```
This does not set the error, it sets the maximum error. Normally this is set based on the maximum velocity and acceleration.

lthall commented on libraries/AC_AttitudeControl/AC_PosControl.h:
```diff
@@ -29,245 +27,253 @@
 
-    /// get_vel_z_error_ratio - returns the proportion of error relative to the maximum request
-    float get_vel_z_control_ratio() const { return constrain_float(_vel_z_control_ratio, 0.0f, 1.0f); }
+    /// stop_pos_xy_stabilisation
```
Extended the comment but I am happy with the name.

lthall commented on libraries/AC_AttitudeControl/AC_PosControl.h:
```diff
@@ -29,245 +27,253 @@
+    void set_pos_error_max_z(float error_min, float error_max) { _p_pos_z.set_error_limits(error_min, error_max); }
+
+    // get_pos_error_max_z_cm - get the maximum vertical position error that will be allowed
+    float get_pos_error_max_z_cm() { return _p_pos_z.get_error_max(); }
+    float get_pos_error_min_z_cm() { return _p_pos_z.get_error_min(); }
```
They are used.

lthall commented on libraries/AC_AttitudeControl/AC_PosControl.h:
```diff
@@ -29,245 +27,253 @@
+    ///     using the default position control kinimatic path.
+    void set_alt_target_with_slew(const float& pos);
+
+    // is_active_z - returns true if the z position controller has been run in the previous 5 loop times
```
I must have missed that conversation. The function still says 5 loop times.

lthall commented on libraries/AC_AttitudeControl/AC_PosControl.h:
```diff
@@ -319,65 +324,53 @@ class AC_PosControl
-            uint16_t reset_desired_vel_to_pos   : 1;    // 1 if we should reset the rate_to_accel_xy step
-            uint16_t reset_accel_to_lean_xy     : 1;    // 1 if we should reset the accel to lean angle step
-            uint16_t reset_rate_to_accel_z      : 1;    // 1 if we should reset the rate_to_accel_z step
             uint16_t vehicle_horiz_vel_override : 1; // 1 if we should use _vehicle_horiz_vel as our velocity process variable for one timestep
```
I will leave this for the programmers to change as this is how it was done before.

lthall commented on libraries/AC_AttitudeControl/AC_PosControl_Sub.cpp:
```diff
@@ -1,101 +1,56 @@
-        // decelerate feed forward to zero
-        _vel_desired.z = constrain_float(0.0f, _vel_desired.z-vel_change_limit, _vel_desired.z+vel_change_limit);
+    // prevent altitude target from breeching altitude limits
+    if (_alt_min < 0 && _alt_min < _alt_max && _alt_max < 100 && _pos_target.z < _alt_min) {
```
I assume you were looking for is_negative(_alt_min)

lthall commented on libraries/AC_PID/AC_PID_2D.cpp:
```diff
@@ -121,24 +121,49 @@ Vector2f AC_PID_2D::update_all(const Vector2f &target, const Vector2f &measureme
+    return _error * _kp;
+}
+
+Vector2f AC_PID_2D::get_i() const
```
return _derivative * _kd;

I think we can only do I right?

lthall replied:
Compile errors meant I could not do this.

peterbarker replied:
OK, so I omitted `const` - it does work like this:

```
-    Vector2f get_i() const;
+    const Vector2f& get_i() const;
```

(because the function is const you can't return a reference to the object without also marking that as `const`.

And yes, can only do I - I misread.

lthall commented on libraries/AC_WPNav/AC_WPNav.cpp:
```diff
@@ -473,16 +464,20 @@ bool AC_WPNav::advance_wp_target_along_track(float dt)
 
     // pass new target to the position controller
-    _pos_control.set_pos_vel_accel_target(target_pos, target_vel, target_accel);
+    _pos_control.set_pos_vel_accel(target_pos, target_vel, target_accel);
```
Lets see those tears!
(I have to think about names some more but math first)

lthall commented on libraries/AP_Math/control.cpp:
```diff
@@ -22,19 +22,190 @@
+
+    float jerk_max = 0.0;
+    if (is_negative(accel_min) && is_positive(accel_max)){
+        jerk_max = MAX(-accel_min, accel_max) * KPa;
```
No, in this case it is the maximum. Here we must ensure that the system is stable for the highest acceleration and therefore the slowest reaction.

lthall commented on libraries/AC_AttitudeControl/AC_PosControl.h:
```diff
@@ -392,28 +385,28 @@ class AC_PosControl
+    float       _tc_z_s;                // time constant of the z kinimatic path generation in seconds used to determine how quickly the aircraft varies the acceleration target
+    float       _vel_max_xy_cms;        // max horizontal speed in cm/s
+    float       _vel_max_up_cms;        // max climb rate in cm/s
+    float       _vel_max_down_cms;      // max descent rate in cm/s
```
I think this makes the variable definitions messy and is different to the way Randy has been doing things so I will leave this up to him.

lthall replied:
one for @rmackay9

lthall commented on libraries/AC_AttitudeControl/AC_PosControl.cpp:
```diff
@@ -271,6 +270,24 @@ const AP_Param::GroupInfo AC_PosControl::var_info[] = {
+    // @Range: 0 10
+    // @Increment: 0.01
+    // @User: Advanced
+    AP_GROUPINFO("_TC_XY", 8, AC_PosControl, _shaping_tc_xy_s, POSCONTROL_DEFAULT_SHAPER_TC),
```
One for Randy

lthall commented on ArduSub/control_surface.cpp:
```diff
@@ -30,6 +28,7 @@ void Sub::surface_run()
         motors.set_desired_spool_state(AP_Motors::DesiredSpoolState::GROUND_IDLE);
         attitude_control.set_throttle_out(0,true,g.throttle_filt);
         attitude_control.relax_attitude_controllers();
+        pos_control.relax_z_controller(motors.get_throttle_hover());
```
This should be just an init

tridge commented on libraries/SITL/SIM_Blimp.cpp:
```diff
@@ -28,6 +28,7 @@ Blimp::Blimp(const char *frame_str) :
 {
     frame_height = 0.0;
     ground_behavior = GROUND_BEHAVIOR_NONE;
+    lock_step_scheduled = true;
```
no, many are not lock-step

tridge replied:
could add tracker, although it doesn't actually matter

peterbarker commented:
On Tue, 18 May 2021, Leonard Hall wrote:

> I don't know how to write ro run unit tests.

https://github.com/peterbarker/ardupilot/pull/new/pr/pcu-unit-test

We have documentation on running them here: 
https://ardupilot.org/dev/docs/unit-tests.html

I was repeatedly running that test I linked to there with this:
./waf build --target=tests/test_sqrt_controller && 
./build/linux/tests/test_sqrt_controller

(after ./waf configure --board=linux)

For fixed mathematical functions unit tests provide a convenient way of 
testing the thing is well-behaved across a variety of inputs.  The big 
advantage is that there's no race conditions (c.f. our autotest suite). 
Just maths.

lthall commented:
>> Commit messages need help; "fixes" really isn't adequate. An outline of the macro-level changes in the extended commit message would be good.
Yep, we can do that with the squash at the end.

>> Method renames should have been pulled out as a separate set of commits on which this was based.
I agree but it is hard to do when I need to rename things and restructure things to work out how to get the changes right and then we change the names again at the end. My plan was to get it all worked out then if I had to go back and do the renaming an reordering on master, make obvious changes then basically copy everything over the top of it to get it back to what we have here. It is hard to justify the extra work given that it is all reviewed by then. Randy feels the same way you do but didn't want to make me do all that work effectively for nothing.

>> Why time-constants in some place and Hz in others?
Depends on use. Decays tend to be time constants. Filters tend to be Hz. Time constants tend to be 0.25 to 10 seconds, Hz tend to be 1 to 100 Hz.

lthall commented on libraries/AC_AttitudeControl/AC_PosControl.cpp:
```diff
@@ -279,333 +296,594 @@ const AP_Param::GroupInfo AC_PosControl::var_info[] = {
+    _p_pos_xy(POSCONTROL_POS_XY_P, dt),
+    _pid_vel_xy(POSCONTROL_VEL_XY_P, POSCONTROL_VEL_XY_I, POSCONTROL_VEL_XY_D, 0.0f, POSCONTROL_VEL_XY_IMAX, POSCONTROL_VEL_XY_FILT_HZ, POSCONTROL_VEL_XY_FILT_D_HZ, dt),
+    _dt(dt),
+    _vel_max_down_cms(POSCONTROL_SPEED_DOWN),
```
one for @rmackay9

tridge commented:
@rmackay9 are you ok with this PR now?
@peterbarker are you ok with the changes?

rmackay9 commented:
@tridge I still need to complete testing on a few more flight modes which I hope to complete on Monday.

Still to be tested: AutoTune, FlowHold, Follow, Guided (pos-vel, attitude),

rmackay9 commented:
I've completed testing of the remaining modes, AutoTune, FlowHold, Follow, Guided (pos-vel, attitude) and I'm OK with this going in now.

I found a small-ish issues with Guided's posvel controller but this issue is resolved in the follow up PR https://github.com/ArduPilot/ardupilot/pull/17399.  I've already discussed this with Leonard and it is not a blcoker.

![posvel-before-after](https://user-images.githubusercontent.com/1498098/119318205-faedfa00-bcb3-11eb-9525-21c072504a2f.png)
